src/element/iframe.ts

```diff
@@ -0,0 +1,197 @@
+    return { link, aspectRatio, type };
+  }
+
+  embeddedLinkCache.set(link, { link, aspectRatio, type });
+  return { link, aspectRatio, type };
+};
+
+export const hideActionForIFrame = (
+  element: ExcalidrawElement | undefined,
+  props: ExcalidrawProps,
+) =>
+  isIFrameElement(element) &&
+  element.link &&
+  element.link !== "" &&
+  !props.iframeURLWhitelist;
+
+export const isIFrameOrFrameLabel = (
+  element: NonDeletedExcalidrawElement,
+): Boolean => {
+  if (isIFrameElement(element)) {
+    return true;
+  }
+  if (element.type === "text") {
+    const container = getContainerElement(element);
+    if (container && isIFrameElement(container)) {
+      return true;
+    }
+  }
+  return false;
+};
+
+export const createPlaceholderiFrameLabel = (
+  element: NonDeletedExcalidrawElement,
+): ExcalidrawElement => {
+  const text =
+    !element.link || element?.link === "" ? "Empty Web-Embed" : element.link;
+  const fontSize = element.width / text.length;
+  return newTextElement({
+    x: element.x + element.width / 2,
+    y: element.y + element.height / 2,
+    strokeColor:
+      element.strokeColor !== "transparent" ? element.strokeColor : "black",
+    backgroundColor: "transparent",
+    fontFamily: FONT_FAMILY.Helvetica,
+    fontSize,
+    text,
+    textAlign: "center",
+    verticalAlign: VERTICAL_ALIGN.MIDDLE,
+    angle: element.angle ?? 0,
+  });
+};
+
+export const actionSetIFrameAsActiveTool = register({
+  name: "setIFrameAsActiveTool",
+  trackEvent: { category: "toolbar" },
+  perform: (elements, appState, _, app) => {
+    const nextActiveTool = updateActiveTool(appState, {
+      type: "iframe",
+    });
+
+    setCursorForShape(app.canvas, {
+      ...appState,
+      activeTool: nextActiveTool,
+    });
+
+    return {
+      elements,
+      appState: {
+        ...appState,
+        activeTool: updateActiveTool(appState, {
+          type: "iframe",
+        }),
+      },
+      commitToHistory: false,
+    };
+  },
+  keyTest: (event) => event.key.toLocaleLowerCase() === KEYS.W,
+});
+
+export class IFrameURLValidator {
+  private static instance: IFrameURLValidator;
+  private validators: RegExp[];
+
+  private constructor(validators: RegExp[] = []) {
+    this.validators = validators;
+  }
+
+  public static getInstance(validators?: RegExp[]): IFrameURLValidator {
+    if (!IFrameURLValidator.instance) {
+      IFrameURLValidator.instance = new IFrameURLValidator(validators);
+    }
+    return IFrameURLValidator.instance;
+  }
+
+  public run(url: string | null | undefined): boolean {
+    if (!url) {
+      return false;
+    }
+    for (const validator of this.validators) {
+      if (url.match(validator)) {
+        return true;
+      }
+    }
+    return Boolean(
+      url.match(YOUTUBE_REG) ||
+        url.match(VIMEO_REG) ||
+        //url.match(TWITTER_REG) ||
+        url.match(FIGMA_REG),
+      //|| url.match(EXCALIDRAW_REG),
+    );
```

Comment:
right now even if you supply a custom validator it will always allow youtube/vimeo/figma, so people wouldn't be able to disable those if needed.

We should either make supplying a custom validator automatically not use our own whitelist (thus forcing people to whitelist every domain they want), or support returning `boolean | undefined` where `boolean` forces the validator to exit early, and `undefined` to use our default lit.

Reply:
What if we'd declare validators like this: `validators: (RegExp | boolean)[];` ? This would have two benefits. To allow all URLs you would only need to pass in `[true]`, if you want to stop validation at the end of your own validators you can finish the array with a `false` e.g.: `[/my first/, /my second/, false]`. If you don't stop the list with false, then the built in validators will follow.

And change run() like this
```typescript
public run(url: string | null | undefined): boolean {
    if (!url) {
      return false;
    }
    for (const validator of this.validators) {
      if (typeof validator === 'boolean') {
        return validator;
      }
      if (url.match(validator)) {
        return true;
      }
    }
    return Boolean(
      url.match(YOUTUBE_REG) ||
        url.match(VIMEO_REG) ||
        url.match(FIGMA_REG),
    );
  }
}

Reply:
I've pushed a commit to implement the above proposed solution